plugins/deployment_manager/DeploymentManager.ts

```diff
@@ -133,46 +146,105 @@ export class DeploymentManager {
+      return contract;
+    }
+    return maybeExisting;
+  }
+
   /* Deploys a contract from Hardhat artifacts */
-  async deploy<
+  async _deploy<
     C extends Contract,
     Factory extends Deployer<C, DeployArgs>,
     DeployArgs extends Array<any>
   >(contractFile: string, deployArgs: DeployArgs, retries?: number): Promise<C> {
-    return await this.asyncCallWithRetry(
+    const contract = await this.retry(
       async () => deploy<C, Factory, DeployArgs>(contractFile, deployArgs, this.hre, await this.deployOpts()),
       retries
     );
+    this.counter++;
+    return contract;
   }
 
   /* Deploys a contract from a build file, e.g. an one imported contract */
-  async deployBuild(buildFile: BuildFile, deployArgs: any[], retries?: number): Promise<Contract> {
-    return await this.asyncCallWithRetry(
+  async _deployBuild<C extends Contract>(buildFile: BuildFile, deployArgs: any[], retries?: number): Promise<C> {
+    const contract = await this.retry(
       async () => deployBuild(buildFile, deployArgs, this.hre, await this.deployOpts()),
       retries
     );
+    this.counter++;
+    return contract;
   }
 
   /* Deploys missing contracts from the deployment, using the user-space deploy.ts script */
-  async deployMissing(force: boolean = false) {
-    // XXX if this is idempotent we can just always deploy, and do the same for dev
-    //  as is, won't handle cases where the deploy script adds roots or partial redeploys
-    // if force or there are no roots, deploy
-    //  force will also have to change with idempotent deploy changes
-    //   its here for deploy task, which doesn't really care if roots exists or not
-    //    but we'll want another way to specify how idempotent should work
-    const roots = await this.getRoots();
-    if (force || roots.size == 0) {
-      // XXX noted above but cache is a misnomer since we have non-cache files its kind of managing
-      //  could either rename or move that functionality
-      const deployScript = this.cache.getFilePath({ rel: 'deploy.ts' });
-      // XXX expect returns roots, and we write them?
-      const { default: deploy } = await import(deployScript);
-      if (!deploy || !deploy.call) {
-        throw new Error(`Missing deploy function in ${deployScript}.`);
-      }
-      await deploy(this);
+  async runDeployScript(deploySpec: object): Promise<DeploymentDelta> {
+    const oldCount = this.counter;
+    const oldSpider = await this.spider();
+    const deployScript = this.cache.getFilePath({ rel: 'deploy.ts' });
+    const { default: deployFn } = await import(deployScript);
+    if (!deployFn || !deployFn.call) {
+      throw new Error(`Missing deploy function in ${deployScript}.`);
     }
+    const deployed = await deployFn(this, deploySpec);
+    const newCount = this.counter;
+    const newSpider = await this.spider(deployed);
+    return {
+      old: { count: oldCount, spider: oldSpider },
```

Comment:
What's the point of tracking the counts?

Reply:
So we can print it out before the diff as an extra sanity check when running deploy from CLI or the initial deploy to each world for the scens, its a message that prints with/without DEBUG so you can eyeball that the number of contracts being deployed makes sense. In general once we have a real deployment there shouldn't be too many contracts ever being deployed